corehq/apps/sms/views.py

```diff
@@ -665,36 +665,110 @@ def global_backend_map(request):
+    data = {}
+    for doc in iter_docs(CommCareCase.get_db(), case_ids):
+        if domain_obj.custom_case_username:
+            name = doc.get(domain_obj.custom_case_username, _('unknown'))
+        else:
+            name = doc.get('name', _('unknown'))
+        data[doc['_id']] = [name]
+    return data
+
+
+def get_mobile_worker_contact_info(domain_obj, user_ids):
+    data = {}
+    for doc in iter_docs(CommCareUser.get_db(), user_ids):
+        user = CommCareUser.wrap(doc)
+        data[user.get_id] = user.raw_username
+    return data
+
+
+def get_contact_info(domain):
+    verified_number_ids = VerifiedNumber.by_domain(domain, ids_only=True)
+    domain_obj = Domain.get_by_name(domain, strict=True)
+    case_ids = []
+    mobile_worker_ids = []
+    data = []
+    for doc in iter_docs(VerifiedNumber.get_db(), verified_number_ids):
+        owner_id = doc['owner_id']
+        if doc['owner_doc_type'] == 'CommCareCase':
+            case_ids.append(owner_id)
+            data.append([
+                None,
+                'case',
+                doc['phone_number'],
+                owner_id,
+            ])
+        elif doc['owner_doc_type'] == 'CommCareUser':
+            mobile_worker_ids.append(owner_id)
+            data.append([
+                None,
+                'mobile_worker',
+                doc['phone_number'],
+                owner_id,
+            ])
+    contact_data = get_case_contact_info(domain_obj, case_ids)
+    contact_data.update(get_mobile_worker_contact_info(domain_obj, mobile_worker_ids))
+    for row in data:
+        contact_info = contact_data.get(row[3])
+        row[0] = contact_info[0] if contact_info else _('(unknown)')
+    return data
+
+
+def format_contact_data(domain, data):
+    for row in data:
+        contact_id = row[3]
+        if row[1] == 'case':
+            row[1] = _('Case')
+            row.append(reverse('case_details', args=[domain, contact_id]))
+        elif row[1] == 'mobile_worker':
+            row[1] = _('Mobile Worker')
+            row.append(reverse(EditCommCareUserView.urlname, args=[domain, contact_id]))
+        else:
+            row.append('#')
+        row.append(reverse('sms_chat', args=[domain, contact_id]))
+
+
+@require_permission(Permissions.edit_data)
+@requires_privilege_with_fallback(privileges.OUTBOUND_SMS)
+def chat_contact_list(request, domain):
+    sEcho = request.GET.get('sEcho')
+    iDisplayStart = int(request.GET.get('iDisplayStart'))
+    iDisplayLength = int(request.GET.get('iDisplayLength'))
+    sSearch = request.GET.get('sSearch', '').strip()
+
+    data = get_contact_info(domain)
+    total_records = len(data)
+
+    if sSearch:
+        regex = re.compile('^.*%s.*$' % sSearch)
+        data = filter(lambda row: regex.match(row[0]) or regex.match(row[2]), data)
+    filtered_records = len(data)
+
+    data.sort(key=lambda row: row[0])
+    data = data[iDisplayStart:iDisplayStart + iDisplayLength]
```

Comment:
Is it not possible to do the paging without loading ALL the data first? Seems that paging this way is pretty pointless and we should rather just send all the data to the browser first time. I see that sorting is also happening in python so guessing that its not possible to page without loading all the data.

Is it maybe possible to at least cache the list so that we don't have to regenerate it every time?